src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs

```diff
@@ -238,38 +240,79 @@ private void OnFileRead(string filePath)
-                sdkResolverContext.Logger.LogMessage(string.Format(CultureInfo.CurrentCulture, Strings.FailedToReadGlobalJson, globalJsonPath, e.Message));
+                try
+                {
+                    json = File.ReadAllText(globalJsonPath);
+                }
+                catch (Exception e)
+                {
+                    // Failed to read file "{0}". {1}
+                    sdkResolverContext.Logger.LogMessage(string.Format(CultureInfo.CurrentCulture, Strings.FailedToReadGlobalJson, globalJsonPath, e.Message));
 
-                return null;
-            }
+                    return null;
+                }
+
+                OnFileRead(globalJsonPath);
+
+                // Look ahead in the contents to see if there is an msbuild-sdks section.  Deserializing the file requires us to load
+                // Newtonsoft.Json which is 500 KB while a global.json is usually ~100 bytes of text.
+                if (json.IndexOf(MSBuildSdksPropertyName, StringComparison.Ordinal) == -1)
+                {
+                    return null;
+                }
 
-            OnFileRead(globalJsonPath);
+                try
+                {
+                    return ParseMSBuildSdkVersionsFromJson(json);
+                }
+                catch (Exception e)
+                {
+                    // Failed to parse "{0}". {1}
+                    sdkResolverContext.Logger.LogMessage(string.Format(CultureInfo.CurrentCulture, Strings.FailedToParseGlobalJson, globalJsonPath, e.Message));
 
-            // Look ahead in the contents to see if there is an msbuild-sdks section.  Deserializing the file requires us to load
-            // Newtonsoft.Json which is 500 KB while a global.json is usually ~100 bytes of text.
-            if (json.IndexOf(MSBuildSdksPropertyName, StringComparison.Ordinal) == -1)
+                    return null;
+                }
+            }
+            finally
             {
-                return null;
+                TraceEvents.GlobalJsonReadStop(globalJsonPath, sdkResolverContext);
             }
+        }
 
-            try
+        private static class TraceEvents
+        {
+            private const string EventNameGlobalJsonRead = "SdkResolver/GlobalJsonRead";
+
+            public static void GlobalJsonReadStart(string globalJsonPath, SdkResolverContext sdkResolverContext)
             {
-                return ParseMSBuildSdkVersionsFromJson(json);
+                var eventOptions = new EventSourceOptions
+                {
+                    ActivityOptions = EventActivityOptions.Detachable,
+                    Keywords = NuGetEventSource.Keywords.SdkResolver | NuGetEventSource.Keywords.Performance,
+                    Opcode = EventOpcode.Start
+                };
+
+                NuGetEventSource.Instance.Write(EventNameGlobalJsonRead, eventOptions, new GlobalJsonReadEventData(globalJsonPath, sdkResolverContext.ProjectFilePath, sdkResolverContext.SolutionFilePath));
```

Comment:
nit: Please consider wrapping the body of these methods in `if (NuGetEventSource.Instance.IsEnabled()`.

The struct initialization here is probably negligible but `Write<T>()` will be JITted for each different event data even if tracing is not enabled and the effect will likely be measurable in some scenarios (e.g. running the resolver in MSBuild.exe where we currently can't use a native image).

Reply:
Thanks, I assumed that `Write<T>()` would just check `IsEnabled()` first?  But you're saying the actual call to `Write<T>()` will potentially cause a perf issue?

Reply:
Yes, the method internally checks `IsEnabled()` but it needs to be JITted first, and since it's generic and instantiated with a different value type in each of these call sites, it will be JITted several times. Not a huge deal .. but easily avoidable.

Reply:
Thank you for making the change!